src/main/java/org/apache/sling/feature/cpconverter/handlers/BundleEntryHandler.java

```diff
@@ -108,111 +141,293 @@ public void handle(@NotNull String path, @NotNull Archive archive, @NotNull Entr
+                    IOUtils.copy(bundleFileInputStream, tmpBundleOutput);
+                }
+            }
+            // TODO: map path to imitate content-package structure
+            SingleFileArchive archive = new SingleFileArchive(tmpInputFile.toFile(), contentPackagePath);
+            entryHandler.handle(repositoryPath, archive, archive.getRoot(), converter);
+            Files.delete(tmpInputFile);
+        } else {
+            // ... otherwise add it to the content package
+            if (tmpInputFile != null) {
+                packageAssembler.addEntry(contentPackagePath, tmpInputFile.toFile());
+                Files.delete(tmpInputFile);
+            } else {
+                packageAssembler.addEntry(contentPackagePath, bundleFileInputStream);
+            }
+        }
+        return true;
+    }
 
-                String epHeader = manifest.getMainAttributes().getValue(Constants.EXPORT_PACKAGE);
-                if (epHeader != null) {
-                    for (Clause clause : Parser.parseHeader(epHeader)) {
-                        converter.getFeaturesManager().addAPIRegionExport(runMode, clause.getName());
+    JcrNamespaceRegistry createNamespaceRegistry(@NotNull Manifest manifest, @NotNull JarFile jarFile) throws RepositoryException, IOException, ParseException {
+        JcrNamespaceRegistry registry = new JcrNamespaceRegistry();
+        // configured CND
+        
+        // parse Sling-Namespaces header (https://github.com/apache/sling-org-apache-sling-jcr-base/blob/66be360910c265473799635fcac0e23895898913/src/main/java/org/apache/sling/jcr/base/internal/loader/Loader.java#L192)
+        final String namespaceDefinition = manifest.getMainAttributes().getValue(NAMESPACES_BUNDLE_HEADER);
+        if (namespaceDefinition != null) {
+            final StringTokenizer st = new StringTokenizer(namespaceDefinition, ",");
+
+            while ( st.hasMoreTokens() ) {
+                final String token = st.nextToken().trim();
+                int pos = token.indexOf('=');
+                if ( pos == -1 ) {
+                    logger.warn("createNamespaceRegistry: Bundle {} has an invalid namespace manifest header entry: {}",
+                            manifest.getMainAttributes().getValue(Constants.BUNDLE_SYMBOLICNAME), token);
+                } else {
+                    final String prefix = token.substring(0, pos).trim();
+                    final String namespace = token.substring(pos+1).trim();
+                    registry.registerNamespace(prefix, namespace);
+                }
+            }
+        }
+        
+        // parse Sling-Nodetypes header
+        final String typesHeader = manifest.getMainAttributes().getValue(NODETYPES_BUNDLE_HEADER);
+        if (typesHeader != null) {
+            StringTokenizer tokener = new StringTokenizer(typesHeader, ",");
```

Comment:
why not using org.apache.sling.commons.osgi.ManifestHeader here?

Reply:
Good idea, will change that. Just copied the code over from https://github.com/apache/sling-org-apache-sling-jcr-base/blob/66be360910c265473799635fcac0e23895898913/src/main/java/org/apache/sling/jcr/base/internal/loader/Loader.java#L192.